### PR TITLE
Upgrade Gradle config

### PR DESCRIPTION
**Pull request by aSemy**
Some prep work for #29 

- [x] update Gradle
- [x] set up buildSrc for shared build config

### CODE REVIEW

**Comment by lgwillmore** [comment]
Give me a shout if you need a hand on the orchid documentation stuff.

**Comment by aSemy** [comment]
> Give me a shout if you need a hand on the orchid documentation stuff.

Thanks! Will do. I was going to try and avoid touching anything that isn't a `.kts` file

I see an error on Netlify. I think it's because Gradle generates Kotlin DSL accessors, and I changed the Gradle config so it also uses Java 11, just like the project itself. Netlify must have cached these DSL accessors. Do you know if it's possible to clear the cache completely?

```
1:20:41 PM:       > gradle/kotlin/dsl/plugins/_5c4ef24a3eefe27a778e93332a6d4eb0/ComPluginGroup has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

**Comment by lgwillmore** [comment]
> > Give me a shout if you need a hand on the orchid documentation stuff.
> 
> Thanks! Will do. I was going to try and avoid touching anything that isn't a `.kts` file
> 
> I see an error on Netlify. I think it's because Gradle generates Kotlin DSL accessors, and I changed the Gradle config so it also uses Java 11, just like the project itself. Netlify must have cached these DSL accessors. Do you know if it's possible to clear the cache completely?
> 
> ```
> 1:20:41 PM:       > gradle/kotlin/dsl/plugins/_5c4ef24a3eefe27a778e93332a6d4eb0/ComPluginGroup has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
> ```

Ya, I think we can ignore the netlify documentation errors for now. I am going to do a cleanup and update of the docs soon. Lets fix the ktlint check though.

**Comment by aSemy** [comment]
> Lets fix the ktlint check though.

Sure thing, that will be easy. I've disabled it for now because it interferes with me testing.

File: core/build.gradle.kts [hunk 1]
```diff
@@ -1,56 +1,42 @@
...
             dependencies {
-                implementation(kotlin("stdlib"))
-                implementation("org.jetbrains.kotlin:kotlin-reflect:$kotlinVersion")
+                implementation(kotlin("reflect"))
             }
         }
-        // JVM-specific tests and their dependencies:
-        jvm().compilations["test"].defaultSourceSet {
+
+        val jvmTest by getting {
```
Comment (aSemy) [inline core/build.gradle.kts:36]: Changing `jvm().compilations["test"].defaultSourceSet` is a unnecessary but I like it if it's consistent with `commonMain` :)

File: core/build.gradle.kts [hunk 2]
```diff
@@ -59,44 +45,23 @@ kotlin {
 publishing {
     publications {
         create<MavenPublication>("coreJVM") {
```
Comment (aSemy) [inline core/build.gradle.kts:47]: Gradle will automatically set the artifactId and version based on the subproject (also in the POM), so I've removed the explicit assignment. If you want to leave it in to be explicit, that's an option too.

File: core/build.gradle.kts [hunk 3]
```diff
@@ -59,44 +45,23 @@ kotlin {
...
-            groupId = "codes.laurence.warden"
-            artifactId = "warden-core"
-            version = version
+            artifact(tasks.jvmJar)
```
Comment (aSemy) [inline core/build.gradle.kts:48]: I set the artifacts to be based on task inputs, so Gradle will automatically be clever and trigger those tasks if and only if required.

File: core/build.gradle.kts [hunk 4]
```diff
@@ -59,44 +45,23 @@ kotlin {
...
-tasks {
-    val build by existing
-
-    artifactoryPublish {
-        dependsOn(build)
-        dependsOn("publishJvmPublicationToMavenLocal")
-    }
+tasks.artifactoryPublish.configure {
+    dependsOn(tasks.build)
+//    dependsOn(tasks.publishJvmPublicationToMavenLocal)
```
Comment (aSemy) [inline core/build.gradle.kts:66]: I think a dependency on `tasks.publishJvmPublicationToMavenLocal` is unnecessary? But maybe there's a reason for it.

**Comment by aSemy** [comment]
Okay, all done! Some usage notes

* I wasn't able to test Artifactory publication because I don't have credentials. I did test publishing it to a local Maven dir, and it seemed to work well.
* `buildSrc/build.gradle.kts` is the singular place where Gradle plugins should be declared as dependencies, and their versions set. In all other plugin blocks, there's no need to include a version.
* It would be nice to use Gradle version catalog to share the versions between `buildSrc` and the root project. https://docs.gradle.org/current/userguide/platforms.html#sec:sharing-catalogs

File: core/build.gradle.kts (hunk 4 above)
Comment (lgwillmore) [reply core/build.gradle.kts:64]: If I remember correctly it was grabbing the output jar from the build directory to push to artifactory, but if we are moving to maven we can scrap the artifactory stuff. Will see how we go when we swap in the sonatype repo in.

**Comment by lgwillmore** [comment]
Cool - looking good on first inspection. Late here, and I am on a plane tomorrow. Want to just go through it in more detail over next couple of days.

**Comment by lgwillmore** [comment]
Update: Just getting group id set up on sonatype

**Review by lgwillmore — APPROVED** [review]
lgtm

**Pull request merged**